File: app/src/main/java/com/example/sy43/viewmodels/CategoryViewModel.java
package com.example.sy43.viewmodels;

import android.util.Log;
import android.widget.Toast;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.sy43.CategoryActivity;
import com.example.sy43.db.entity.Categorydb;
import com.example.sy43.models.Category;
import com.example.sy43.repositories.CategoryRepository;

import java.util.List;
import java.util.stream.Collectors;

public class CategoryViewModel extends ViewModel {
    private MutableLiveData<List<Categorydb>> categories;
    private MutableLiveData<List<Categorydb>> objectives;

    public CategoryRepository catRepo;

    public void init() {
        if (categories != null) {
            return;
        }

        catRepo = CategoryRepository.getInstance();
        categories = catRepo.getCategories();
        objectives = catRepo.getObjectives();

    }

    public void createCategory(final Categorydb category) {
        this.catRepo.createCategory(category);
        this.categories.postValue(this.categories.getValue());
    }

    public void delCategories(int id) {
        this.catRepo.deleteCategory(id);
    }

    public LiveData<Categorydb> getCategoryById(int id){
        return this.catRepo.getCategoryById(id);
    }
    public LiveData<List<Categorydb>> getCategories() {
        return categories;
    }
    public LiveData<List<Categorydb>> getObjectives() {
        return objectives;
    }
}

File: app/src/main/java/com/example/lafo_cheuse/viewmodels/CategoryViewModel.kt
package com.example.lafo_cheuse.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import com.example.lafo_cheuse.models.Category
import com.example.lafo_cheuse.models.Income
import com.example.lafo_cheuse.repository.CategoryRepository
import com.example.lafo_cheuse.repository.IncomeRepository
import kotlinx.coroutines.launch

/**
 * ViewModel of the categories.
 *
 * @property repository - An [IncomeRepository] instance which will receive the queries
 * @property allCategories - A [LiveData] list of all the [Category]
 *
 * @constructor It will initialize all the properties with the [LiveData] asked from the [repository]
 *
 * @param application - The [Application] which it is linked with
 */
class CategoryViewModel(application: Application) : AndroidViewModel(application) {
    private var repository : CategoryRepository
    var allCategories : LiveData<List<Category>>?

    init {
        repository = CategoryRepository(application)
        allCategories = repository.getCategories()
    }

    /**
     * Method to get all the categories
     *
     * @return [allCategories]
     */
    fun getCategories() : LiveData<List<Category>>? {
        return allCategories
    }

    /**
     * Method to get a particular category
     *
     * @param categoryName - a [String] matching with the name of a [Category]
     * @param categoryEmoji - a [String] matching with the emoji of a [Category]
     * @return a [LiveData] list of [Category]
     */
    fun getCategory(categoryName : String, categoryEmoji : String) : LiveData<List<Category>>? {
        return repository.getCategory(categoryName,categoryEmoji)
    }

    /**
     * Method to get a particular category synchronously
     *
     * @param categoryName - a [String] matching with the name of a [Category]
     * @param categoryEmoji - a [String] matching with the emoji of a [Category]
     * @return a list of [Category]
     */
    fun getCategorySync(categoryName : String, categoryEmoji : String): List<Category>? {
        return repository.getCategorySync(categoryName,categoryEmoji)
    }

    /**
     * Method to return the default category ( the extras category)
     *
     * @return a [LiveData] list of [Category] only containing one [Category]
     */
    fun getDefaultCategory() : LiveData<List<Category>>? {
        return getCategory("extras","❔")
    }

    /**
     * Method to insert a [Category] in the database
     *
     * @param category - [Category] to insert in the database
     * @return [Unit] nothing
     */
    fun createCategory(category: Category) = viewModelScope.launch {
        repository.createCategory(category)
    }

    /**
     * Method to update a [Category] in the database
     *
     * @param category - [Category] to update
     * @return [Unit] nothing
     */
    fun updateCategory(category: Category) = viewModelScope.launch {
        repository.updateCategory(category)
    }

    /**
     * Method to delete a [Category] from the database
     *
     * @param bId - ID of the [Category] to delete
     * @return [Unit] nothing
     */
    fun deleteCategory(bId : Long) = viewModelScope.launch {
        repository.deleteCategory(bId)
    }

    /**
     * Method to delete all [Category] from the database
     *
     * @return [Unit] nothing
     */
    fun deleteAllCategories() = viewModelScope.launch {
        repository.deleteAllCategories()
    }
}